### ISSUE TITLE
Turn off ASLR if possible

### ISSUE THREAD

**Issue by vext01** [softdevteam/krun#70]
We should turn off address space layout randomisation on platforms that support it.

**Comment by ltratt** [comment]
It's unclear that this is a good idea (for the same reasons that fixing the environment size may not be a good idea).

**Comment by vext01** [comment]
Can you expand upon that? Not saying you are wrong, but as it stands, I don't understand. Surely anything using randomness is a uncontrolled variable?

I think @snim2 had a draft paper that proposed this? What do you think Sarah?

**Comment by snim2** [comment]
@ltratt is right that the two things go together (i.e. they attempt to control the placement of the stack, so that it ends up in the same place on each execution).

There are various ways to turn off ASLR, such as:

```bash
$ echo 0 | sudo tee /proc/sys/kernel/randomize_va_space
```

**Comment by vext01** [comment]
Well, I guess it depends which kind of ASLR we are talking about, as to whether the stack is affected.

Some OSs only load the libraries in a random order, some do that *and* randomise the position of the `.text` section too. For these two schemes I'm not sure the stack is affected as it usually grows down (on common architectures at least) from the top of the memory space and your libs and `.text` section are separate, further down the address space.

Some OSs put random gaps in the stack between frames. Obviously this would affect the stack.

But, I'm still not sure why we wouldn't want to control as many of these variables as possible. Perhaps I am missing something?

**Comment by ltratt** [comment]
The Stabilizer paper gives a good argument as to why fixing these things is not necessarily a good idea.

**Comment by ltratt** [comment]
OK, the consensus is that for the first run we will turn ASLR *off*.

**Comment by vext01** [comment]
I'm going to add the ability for krun to check ASLR is off then.

**Comment by vext01** [comment]
Another note related to ASLR, I supose we should also check that the C programs are not built with `-fPIE` (positional independant executable). when used, this loads `.text` at a random offset in the memory image, although I have no idea its effect when `randomize_va_space` is off.

**Comment by vext01** [comment]
I imagine the section is loaded at a static offset, but that the code still reserves a register to refer to the section. In other words `-fPIE` will contribute to register contention regardless of the system's ASLR settings... Will need to think more if that is an issue...

**Comment by ltratt** [comment]
Good point -- it's worth investigating.

**Comment by vext01** [comment]
I've also just found a compile-time kernel switch "Disable heap randomization". Will also need investigating.

**Comment by vext01** [comment]
FWIW, that is *off* by default.

**Comment by ltratt** [comment]
OK, let's leave it off (for the same reasons that we're leaving ASLR off for the initial run).

**Comment by vext01** [comment]
To be clear, when enabled, the switch *disables* heap randomization. If we leave it off, we enable a ASLR feature.

I need to read into how this interacts with `/proc/sys/kernel/randomize_va_space`. Hopefully this switch turns off heap randomisation regardless...

**Comment by ltratt** [comment]
Ah, yes, Sarah and I encountered randomize_va_space yesterday. I believe we should switch it off, at least for the initial run.

**Comment by vext01** [comment]
OK, I've done some experiments.

With `/proc/sys/kernel/randomize_va_space` set to 2:

 * A binary compiled *without* `-pie -fPIE` will *not* have it's `.text` section randomised. I.e. the address of `main` is fixed across runs of the binary.
 * A binary compiled *with* `-pie -fPIE` *will* have it's `.text` section randomised *and* lookups of various kinds of data are relative to `RIP` or `EIP` (the 64 and 32 bit program counter respectively). For example, the binary I was looking at had addressed a `printf` format string via `[RIP+offset]`.

If you flip off `/proc/sys/kernel/randomize_va_space`, i.e. set it to 0:

* A binary compiled *without* `-pie -fPIE` will *not* have it's `.text` section randomised.
* A binary compiled *with* `-pie -fPIE` will still use `RIP` and `EIP` relative addressing, but the text section will always be relocated to the same place, oddly much higer in the address space than usual. My test binary's main was at `0x5555555547a0` every time.

In summary, PIE is *mostly* independent of `/proc/sys/kernel/randomize_va_space`.

**Comment by ltratt** [comment]
OK, thanks.

**Comment by vext01** [comment]
Regarding the kernel compile-time switch "Disable heap randomization" (aka `CONFIG_COMPAT_BRK`), this just changes the default value of `/proc/sys/kernel/randomize_va_space` to 0.

So it is sufficient to simply check that `/proc/sys/kernel/randomize_va_space` is 0 (i.e. we don't need to check the kernel config like we did for tickless mode)

**Comment by vext01** [comment]
For more info see `CONFIG_COMPAT_BRK` in https://www.kernel.org/doc/Documentation/sysctl/kernel.txt
